### ISSUE TITLE
Exception after 3.2 upgrade

### ISSUE THREAD

**Issue by Padam87** [symfony/symfony#20698]
Hi,

It seems there is a problem with form serialization in 3.2.

```
Exception in FormDataCollector.php line 274: Serialization of 'Closure' is not allowed
```

Disabling the profiler or downgrading to `"symfony/form": "3.1.*"` "fixes" the problem.

https://gist.github.com/Padam87/dfbfd4a63064b0ac45068ad06bcc4876

**Comment by ogizanagi** [comment]
@Padam87 : Could you fork the standard edition with the minimum changes to reproduce this issue?

**Comment by Padam87** [comment]
I will try, but the project I updated has a lot of dependencies, I was hoping someone else has ran into this.

**Comment by Padam87** [comment]
After some digging I found out that I only get these errors on forms rendered in a subrquest.

Submitted a PR which fixes the problem, but IDK if it is the preferred way to solve this.

**Comment by PhilETaylor** [comment]
I get the same Exception error in a different place...

vendor/symfony/symfony/src/Symfony/Component/HttpKernel/DataCollector/DataCollector.php
 line 49

gives 

`Serialization of 'Closure' is not allowed`

This is something to do with `SplObjectStorage->serialize()` from the backtrace - but is beyond me to resolve, for now I have had to revert to 3.1.7

The problem is intermittent on some pages, every page load on some, and absent on others :( 

`Disabling the profiler or downgrading to "symfony/form": "3.1.*" "fixes" the problem.`

Yup that fixed it for me too :(

**Comment by ogizanagi** [comment]
@PhilETaylor : Could you show us a stacktrace?

**Comment by Padam87** [comment]
@PhilETaylor Hi. Is your form created via a subrequest?

**Comment by PhilETaylor** [comment]
@ogizanagi Stacktrace: https://gist.github.com/d361c5b86f03b4e75e8a57463fb8a830

@Padam87 I dont believe this is form based for me, certainly it is broken on pages with several ajax requests on, more than just a basic login page form page.

I have quite a few things in my project including mysql doctrine & redis and I lost the will to dig too deep :(

**Comment by snpy** [comment]
I've been struggling today with this issue and finally after stumbling on this Issue I've decided to create simplest possible case where this appears.
Turns out easiest way to reproduce this it to use twice the same sub request (with form).
Form type can be bare (nothing but interface requirements) and we're getting serialization issue.

Below I'm adding output generated by `git format-patch` command.
These are the only changes made on file set created using `composer create-project symfony/framework-standard-edition` command.
`symfony/symfony` references https://github.com/symfony/symfony/commit/b96a144bc875684f3338f697687147a2696d73eb

```
From 9d747117780dea65a7c080472c754c8561d6b190 Mon Sep 17 00:00:00 2001
From: Marek Snopkowski <snpy@snpy.info>
Date: Fri, 2 Dec 2016 00:34:11 +0100
Subject: [PATCH] Simplest Closure Exception reproduction

---
 src/AppBundle/Controller/DefaultController.php        | 16 ++++++++++------
 src/AppBundle/Form/SimplestType.php                   | 18 ++++++++++++++++++
 src/AppBundle/Resources/views/Default/index.html.twig |  2 ++
 src/AppBundle/Resources/views/Default/sub.html.twig   |  1 +
 4 files changed, 31 insertions(+), 6 deletions(-)
 create mode 100644 src/AppBundle/Form/SimplestType.php
 create mode 100644 src/AppBundle/Resources/views/Default/index.html.twig
 create mode 100644 src/AppBundle/Resources/views/Default/sub.html.twig

diff --git a/src/AppBundle/Controller/DefaultController.php b/src/AppBundle/Controller/DefaultController.php
index 5216afe..78798d0 100644
--- a/src/AppBundle/Controller/DefaultController.php
+++ b/src/AppBundle/Controller/DefaultController.php
@@ -2,20 +2,24 @@

 namespace AppBundle\Controller;

-use Sensio\Bundle\FrameworkExtraBundle\Configuration\Route;
+use AppBundle\Form\SimplestType;
+use Sensio\Bundle\FrameworkExtraBundle\Configuration as View;
 use Symfony\Bundle\FrameworkBundle\Controller\Controller;
 use Symfony\Component\HttpFoundation\Request;

 class DefaultController extends Controller
 {
     /**
-     * @Route("/", name="homepage")
+     * @View\Route("/", name="homepage")
+     * @View\Template
      */
     public function indexAction(Request $request)
     {
-        // replace this example code with whatever you need
-        return $this->render('default/index.html.twig', [
-            'base_dir' => realpath($this->getParameter('kernel.root_dir').'/..').DIRECTORY_SEPARATOR,
-        ]);
+    }
+
+    /** @View\Template */
+    public function subAction()
+    {
+        return ['form' => $this->createForm(SimplestType::class, null, ['action' => ''])->createView()];
     }
 }
diff --git a/src/AppBundle/Form/SimplestType.php b/src/AppBundle/Form/SimplestType.php
new file mode 100644
index 0000000..f586738
--- /dev/null
+++ b/src/AppBundle/Form/SimplestType.php
@@ -0,0 +1,18 @@
+<?php
+
+namespace AppBundle\Form;
+
+use Symfony\Component\Form\AbstractType;
+use Symfony\Component\Form\FormBuilderInterface;
+use Symfony\Component\OptionsResolver\OptionsResolver;
+
+class SimplestType extends AbstractType
+{
+    public function buildForm(FormBuilderInterface $builder, array $options)
+    {
+    }
+
+    public function configureOptions(OptionsResolver $resolver)
+    {
+    }
+}
diff --git a/src/AppBundle/Resources/views/Default/index.html.twig b/src/AppBundle/Resources/views/Default/index.html.twig
new file mode 100644
index 0000000..0a0160c
--- /dev/null
+++ b/src/AppBundle/Resources/views/Default/index.html.twig
@@ -0,0 +1,2 @@
+{{ render(controller('AppBundle:Default:sub')) }}
+{{ render(controller('AppBundle:Default:sub')) }}
diff --git a/src/AppBundle/Resources/views/Default/sub.html.twig b/src/AppBundle/Resources/views/Default/sub.html.twig
new file mode 100644
index 0000000..0c1a4ce
--- /dev/null
+++ b/src/AppBundle/Resources/views/Default/sub.html.twig
@@ -0,0 +1 @@
+{{ form(form) }}
--
2.10.2
```

Finally adding `composer.lock` in case other dependencies could clarify things further.

[composer.lock.txt](https://github.com/symfony/symfony/files/626008/composer.lock.txt)

**Comment by ogizanagi** [comment]
Great. I can reproduce it thanks to your patch 🎉 

So the bug exists when rendering two forms with the same name in 2 or more subrequests.
Given that you may generate some form from your controller with the form builder without explicitely giving it a name (so defaults to `form`), this issue may be quite common and annoying.

Status: Reviewed

**Comment by PhilETaylor** [comment]
Ok some more information for you. I have worked back through the releases to see which releases would work

If my composer.json has `"symfony/symfony": "v3.2.0-BETA1",` then the profiler is working perfectly

If my composer.json has `"symfony/symfony": "v3.2.0-RC1",` then the profiler is broken most of the time

it works at commit 0896b02be073b8fa60fe6e8b3cf5d4f81b748bf8 which is the last one on this page https://github.com/symfony/symfony/commits/3.2?after=XneqwOaaHq3GBuHarRoGvPh3uI8rMzQ%3D

By the time you get to 5e19c51bbdc01221c2237da5f10b2e8d8eb34326 half way up that page its broken.

I can replicate this over and over again in my live app.

**Comment by nicolas-grekas** [comment]
Does #20707 fix the issue for all or some of you?

**Comment by snpy** [comment]
#20707 worked for me (in case presented above).

**Comment by PhilETaylor** [comment]
nope #20707 never worked for me, I still get

```
Exception in DataCollector.php line 49:
Serialization of 'Closure' is not allowed
```

where the toolbar should be in the footer

**Comment by ogizanagi** [comment]
#20707 only fixes the issue for the `FormDataCollector`.

I think the issue reported by @PhilETaylor is a totally different one.

**Comment by PhilETaylor** [comment]
Totally different, yet the same final error and the same head-banging-on-desk annoying result :)

**Comment by SkyAgency** [comment]
Same error: Exception in FormDataCollector.php line 274: Serialization of 'Closure' is not allowed, on 3.1.x it works.

**Comment by nicolas-grekas** [comment]
Does #20700 fix your issue @PhilETaylor ?

**Comment by PhilETaylor** [comment]
@nicolas-grekas Yes, I believe #20700 has resolved my issue I was having after I manually patched the files, the problem I had has been resolved.

**Comment by nicolas-grekas** [comment]
Could anyone provide a reproducer using the standard edition and share it here?

**Comment by jzawadzki** [comment]
@nicolas-grekas here you go: https://github.com/jzawadzki/symfony-standard/commit/1235dfde8cb72ca557d92f93f38092f6fedbb961
It's in situation when you put object with Closure as translation argument.
I can also confirm that fix #20700 resolves it.

**Comment by nicolas-grekas** [comment]
For translation it's OK. What about forms? Reproducer that fails without https://github.com/symfony/symfony/pull/20707?

**Comment by snpy** [comment]
@nicolas-grekas https://github.com/symfony/symfony/issues/20698#issuecomment-264331542 contains patch that applied to standard-edition will reproduce the error.

**Comment by jzawadzki** [comment]
yeap - confirming that patch from @snpy reproduces the error. 
Just in case - here is fork with patch applied: https://github.com/jzawadzki/symfony-standard/commit/a28366eea8a2d4f2934ff8eac3ce7ba1650869ee

**Comment by ogizanagi** [comment]
A minimalistic patch:

```diff
diff --git a/src/AppBundle/Controller/DefaultController.php b/src/AppBundle/Controller/DefaultController.php
index 5216afe..bfc0469 100644
--- a/src/AppBundle/Controller/DefaultController.php
+++ b/src/AppBundle/Controller/DefaultController.php
@@ -4,18 +4,28 @@ namespace AppBundle\Controller;
 
 use Sensio\Bundle\FrameworkExtraBundle\Configuration\Route;
 use Symfony\Bundle\FrameworkBundle\Controller\Controller;
-use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\Response;
 
 class DefaultController extends Controller
 {
     /**
      * @Route("/", name="homepage")
      */
-    public function indexAction(Request $request)
+    public function indexAction()
     {
-        // replace this example code with whatever you need
-        return $this->render('default/index.html.twig', [
-            'base_dir' => realpath($this->getParameter('kernel.root_dir').'/..').DIRECTORY_SEPARATOR,
-        ]);
+        $this->forward('AppBundle:Default:sub');
+        $this->forward('AppBundle:Default:secondSub');
+
+        return Response::create();
+    }
+
+    public function subAction()
+    {
+        $this->get('form.factory')->createNamedBuilder('foo')->add('bar')->getForm()->createView();
+    }
+
+    public function secondSubAction()
+    {
+        $this->get('form.factory')->createNamedBuilder('foo')->add('bar')->getForm()->createView();
     }
 }
```

**Comment by nicolas-grekas** [comment]
Thanks for the reproducers, I'm now able to debug that issue. I'm going to check asap.

**Comment by nicolas-grekas** [comment]
Could you please try #20762? Does it fix the issue?

**Comment by Padam87** [comment]
@nicolas-grekas It is working in my case :) Thank you!

**Comment by viterete8** [comment]
I was having the same issue after upgrading to 3.2.* from composer (the last symfony version was 3.1.7), https://github.com/Padam87/symfony/pull/1/commits/d706187b07e340558fcf09950bd88f46ea1e1d5f fixed my problem.

**Comment by nicolas-grekas** [comment]
@viterete8 this issue is closed because it should be fixed in 3.2-dev at least. Can you please elaborate on your feedback? Can you reproduce with the latest code?

**Comment by viterete8** [comment]
@nicolas-grekas i did the update today, in composer "symfony/symfony": ">=3.2.1" (previously i had 3.1.7) , and i got "Exception in FormDataCollector.php line 272: Serialization of 'Closure' is not allowed" just at the bottom of the page, everything above loaded well but the profile bar didn't show. Just for the record, the 272 line was "return serialize($this->data);". I did the change to Padam87 code and everything worked well after.

**Comment by nicolas-grekas** [comment]
>= 3.2.1 means you may have 3.2.1. But if the fix is in 3.2.next, you won't have it.
Can you please apply https://github.com/symfony/symfony/commit/63087e515ed4f4f9201f5b5843883bb667505227

**Comment by nicolas-grekas** [comment]
Looks like the patch is in 3.2.1. Could you then please provide a reproducer, eg by forking the standard edition first ?
